static/templyte/src/app/download-template/download-template.component.ts

```diff
@@ -64,6 +74,43 @@ export class DownloadTemplateComponent implements OnInit {
         this.authService.logout(null);
     }
 
+    storeRating(new_rating) {
+        const component = this;
+        console.log("new rating: "+new_rating);
+        let authorUID:string  =  component.user.uid;
+        component.templateRatingsInfoDatabaseRef.once('value').then(snapshot2 => {
+            const old_rating = snapshot2.child(authorUID).val(); //value of previous rating
+            console.log("old_rating: "+old_rating);
+            let hasRated:boolean = old_rating != null;
+            if (hasRated == true) {
+                component.templateDirectoryInfoDatabaseRef.child('/numberRatings').transaction(function(numberRatings){
+                    return numberRatings - 1;
+                });
+                component.templateDirectoryInfoDatabaseRef.child('/ratingSum').transaction(function(ratingSum){
+                    return ratingSum - old_rating;
+                });
+            }
+            component.templateRatingsInfoRef.set({
+                [authorUID]: new_rating
+            });
+
+            let ratingSum1, ratingCount //was there before
+            component.templateDirectoryInfoDatabaseRef.child('/ratingSum').transaction(function(ratingSum){
+                return ratingSum + new_rating;
+            })
+            component.templateDirectoryInfoDatabaseRef.child('/numberRatings').transaction(function(numberRatings){
+                return numberRatings + 1;
+            });
+            component.templateDirectoryInfoDatabaseRef.once('value').then(dirInfoSnapshot => {
```

Comment:
Potential single user race condition - what if the read occurs before the rating sum has been updated?

Reply:
Transactions are asynchronous, so this is a problem that could happen. It could be avoided by using update like I was doing before. Should we revert back? If not how do you suggest we address this?

Reply:
You should deal with this once a transaction has finished - within a callback